File: codigo/1D/EF_barra_n_nodos/funciones_forma_lagrangianos1D.py
# -*- coding: utf-8 -*-

import numpy as np
import sympy as sp
from sympy.polys.polyfuncs import interpolate
sp.init_printing(pretty_print=True)

# Nota: el módulo sympy provee unas funciones muy limitadas para la graficación
# Lo mejor será pasar las funciones de forma N_i por sp.lambdify() y graficarlas
# con matplotlib

# %% -------------------------------------------------------------------------
# Funciones de Forma Lagrangianas de dos nodos

# Calculo las funciones de forma
xi = sp.symbols('xi')
N1 = interpolate([(-1,1), (1,0)], xi)
N2 = interpolate([(-1,0), (1,1)], xi)

# Imprimo las funciones de forma
print('\n\nFunciones de Forma Lagrangianas de DOS nodos:')
print(f'\nN1({sp.pretty(xi)}) ='); sp.pprint(N1)
print(f'\nN2({sp.pretty(xi)}) ='); sp.pprint(N2)

# Grafico las funciones de forma
p = sp.plot((N1, (xi, -1, 1)),
            (N2, (xi, -1, 1)), show=False)
p[0].line_color = 'red'
p[0].label = r'$N_1(\xi)$'

p[1].line_color = 'blue'
p[1].label = r'$N_2(\xi)$'

p.xlabel = r'$\xi$'
p.ylabel = ' '
p.title  = 'Funciones de Forma Lagrangianas de DOS nodos'
p.legend = True
p.show()

# %% -------------------------------------------------------------------------
# Funciones de Forma Lagrangianas de tres nodos

# Calculo las funciones de forma
xi = sp.symbols('xi')
N1 = interpolate([(-1,1), (0,0), (1,0)], xi)
N2 = interpolate([(-1,0), (0,1), (1,0)], xi)
N3 = interpolate([(-1,0), (0,0), (1,1)], xi)

# Imprimo las funciones de forma
print('\n\nFunciones de Forma Lagrangianas de TRES nodos:')
print(f'\nN1({sp.pretty(xi)}) ='); sp.pprint(N1)
print(f'\nN2({sp.pretty(xi)}) ='); sp.pprint(N2)
print(f'\nN3({sp.pretty(xi)}) ='); sp.pprint(N3)

# Grafico las funciones de forma
p = sp.plot((N1, (xi, -1, 1)),
            (N2, (xi, -1, 1)),
            (N3, (xi, -1, 1)), show=False)
p[0].line_color = 'red'
p[0].label = r'$N_1(\xi)$'

p[1].line_color = 'blue'
p[1].label = r'$N_2(\xi)$'

p[2].line_color = 'cyan'
p[2].label = r'$N_3(\xi)$'

p.xlabel = r'$\xi$'
p.ylabel = ' '
p.title  = 'Funciones de Forma Lagrangianas de TRES nodos'
p.legend = True
p.show()

# %% -------------------------------------------------------------------------
# Funciones de Forma Lagrangianas de cuatro nodos

# Calculo las funciones de forma
xi = sp.symbols('xi')
N1 = interpolate([(-1,1), (-1/3,0), (1/3,0), (1,0)], xi)
N2 = interpolate([(-1,0), (-1/3,1), (1/3,0), (1,0)], xi)
N3 = interpolate([(-1,0), (-1/3,0), (1/3,1), (1,0)], xi)
N4 = interpolate([(-1,0), (-1/3,0), (1/3,0), (1,1)], xi)

# Imprimo las funciones de forma
print('\n\nFunciones de Forma Lagrangianas de CUATRO nodos:')
print(f'\nN1({sp.pretty(xi)}) ='); sp.pprint(N1)
print(f'\nN2({sp.pretty(xi)}) ='); sp.pprint(N2)
print(f'\nN3({sp.pretty(xi)}) ='); sp.pprint(N3)
print(f'\nN4({sp.pretty(xi)}) ='); sp.pprint(N4)

# Grafico las funciones de forma
p = sp.plot((N1, (xi, -1, 1)),
            (N2, (xi, -1, 1)),
            (N3, (xi, -1, 1)),
            (N4, (xi, -1, 1)), show=False)
p[0].line_color = 'red'
p[0].label = r'$N_1(\xi)$'

p[1].line_color = 'blue'
p[1].label = r'$N_2(\xi)$'

p[2].line_color = 'cyan'
p[2].label = r'$N_3(\xi)$'

p[3].line_color = 'magenta'
p[3].label = r'$N_4(\xi)$'

p.xlabel = r'$\xi$'
p.ylabel = ' '
p.title  = 'Funciones de Forma Lagrangianas de CUATRO nodos'
p.legend = True
p.show()

# %% -------------------------------------------------------------------------
# Funciones de Forma Lagrangianas de cinco nodos

# Calculo las funciones de forma
xi = sp.symbols('xi')
N = 5*[None]
for i in range(5):
    coef = np.polyfit([-1, -1/2, 0, 1/2, 1], [i==0, i==1, i==2, i==3, i==4], 4)
    coef[abs(coef) < 1e-7] = 0 # remueva los coeficientes demasiado pequeños
    N[i] = sp.Poly(coef, xi).as_expr()      

# Imprimo las funciones de forma
print('\n\nFunciones de Forma Lagrangianas de CINCO nodos:\n')
for i in range(5):
    print(f'\n\nN{i+1}({sp.pretty(xi)}) ='); sp.pprint(N[i])

# Grafico las funciones de forma
p = sp.plot(*[(N[i], (xi, -1, 1)) for i in range(5)], show=False)

line_color = [ 'red', 'blue', 'cyan', 'magenta', 'black' ]
for i in range(5):
    p[i].line_color = line_color[i]
    p[i].label = f'$N_{i+1}(\\xi)$'
           
p.xlabel = r'$\xi$'
p.ylabel = ' '
p.title  = 'Funciones de Forma Lagrangianas de CINCO nodos'
p.legend = True
p.show()

File: codigo/1D/EF_barra_n_nodos/funciones_forma_lagrangianos1D.m
clear, clc, close all
%% -------------------------------------------------------------------------
%% Funciones de Forma Lagrangianas de dos nodos

% Calculo las funciones de forma
syms xi
N1 = calc_N([-1 1], [1 0], xi);
N2 = calc_N([-1 1], [0 1], xi);

% Imprimo las funciones de forma
fprintf('\n\nFunciones de Forma Lagrangianas de DOS nodos:\n')
fprintf('\n\nN1(xi) = \n'), pretty(N1)
fprintf('\n\nN2(xi) = \n'), pretty(N2)

% Grafico las funciones de forma
figure                 % Creo un lienzo
grid on                % creo la rejilla
hold on;               % Para que no se sobreescriban los graficos
h1 = fplot(N1, [-1 1], 'Color', 'r', 'LineWidth', 2);
h2 = fplot(N2, [-1 1], 'Color', 'b', 'LineWidth', 2);
title('Funciones de Forma Lagrangianas de DOS nodos')
xlabel('\xi'); 
ylabel('N_i(\xi)');
plot([-1 1],[0 0], 'ko', [-1 1],[1 1], 'ko') % grafico los nodos
legend([h1, h2], 'N1(\xi)','N2(\xi)','Location','Best');

%% -------------------------------------------------------------------------
%% Funciones de Forma Lagrangianas de tres nodos

% Calculo las funciones de forma
syms xi
N1 = calc_N([-1 0 1], [1 0 0], xi); % = xi*(xi-1)/2
N2 = calc_N([-1 0 1], [0 1 0], xi); % = (1+xi)*(1-xi)
N3 = calc_N([-1 0 1], [0 0 1], xi); % = xi*(xi+1)/2

% Imprimo las funciones de forma
fprintf('\n\nFunciones de Forma Lagrangianas de TRES nodos:\n')
fprintf('\n\nN1(xi) = \n'), pretty(N1)
fprintf('\n\nN2(xi) = \n'), pretty(N2)
fprintf('\n\nN3(xi) = \n'), pretty(N3)

% Grafico las funciones de forma
figure                 % Creo un lienzo
grid on                % creo la rejilla
hold on;               % Para que no se sobreescriban los graficos
h1 = fplot(N1, [-1 1], 'Color', 'r', 'LineWidth', 2);
h2 = fplot(N2, [-1 1], 'Color', 'b', 'LineWidth', 2);
h3 = fplot(N3, [-1 1], 'Color', 'c', 'LineWidth', 2);
title('Funciones de Forma Lagrangianas de TRES nodos')
xlabel('\xi'); 
ylabel('N_i(\xi)');
plot([-1 0 1],[0 0 0], 'ko', [-1 0 1],[1 1 1], 'ko') % grafico los nodos
legend([h1, h2, h3], 'N1(\xi)','N2(\xi)','N3(\xi)','Location','Best');

%% -------------------------------------------------------------------------
%% Funciones de Forma Lagrangianas de cuatro nodos

% Calculo las funciones de forma
syms xi
N1 = calc_N([-1 -1/3 1/3 1], [1 0 0 0], xi);
N2 = calc_N([-1 -1/3 1/3 1], [0 1 0 0], xi);
N3 = calc_N([-1 -1/3 1/3 1], [0 0 1 0], xi);
N4 = calc_N([-1 -1/3 1/3 1], [0 0 0 1], xi);

% Imprimo las funciones de forma
fprintf('\n\nFunciones de Forma Lagrangianas de CUATRO nodos:\n')
fprintf('\n\nN1(xi) = \n'), pretty(N1);  fprintf('\n\nN2(xi) = \n'), pretty(N2)
fprintf('\n\nN3(xi) = \n'), pretty(N3);  fprintf('\n\nN4(xi) = \n'), pretty(N4)

% Grafico las funciones de forma
figure                 % Creo un lienzo
grid on                % creo la rejilla
hold on;               % Para que no se sobreescriban los graficos
h1 = fplot(N1, [-1 1], 'Color', 'r', 'LineWidth', 2);
h2 = fplot(N2, [-1 1], 'Color', 'b', 'LineWidth', 2);
h3 = fplot(N3, [-1 1], 'Color', 'c', 'LineWidth', 2);
h4 = fplot(N4, [-1 1], 'Color', 'm', 'LineWidth', 2);
title('Funciones de Forma Lagrangianas de CUATRO nodos')
xlabel('\xi'); 
ylabel('N_i(\xi)');
plot([-1 -1/3 1/3 1],[0 0 0 0], 'ko', [-1 -1/3 1/3 1],[1 1 1 1], 'ko')
axis([-1 1 -0.4 1.2])
legend([h1 h2 h3 h4], 'N1(\xi)','N2(\xi)','N3(\xi)','N4(\xi)','Location','Best');

%% -------------------------------------------------------------------------
%% Funciones de Forma Lagrangianas de cinco nodos

% Calculo las funciones de forma
syms xi
N = cell(5,1);
N{1} = calc_N([-1 -1/2 0 1/2 1], [1 0 0 0 0], xi);
N{2} = calc_N([-1 -1/2 0 1/2 1], [0 1 0 0 0], xi);
N{3} = calc_N([-1 -1/2 0 1/2 1], [0 0 1 0 0], xi);
N{4} = calc_N([-1 -1/2 0 1/2 1], [0 0 0 1 0], xi);
N{5} = calc_N([-1 -1/2 0 1/2 1], [0 0 0 0 1], xi);

% Imprimo las funciones de forma
fprintf('\n\nFunciones de Forma Lagrangianas de CINCO nodos:\n')
for i = 1:5
    fprintf('\n\nN%d(xi) = \n', i); pretty(N{i});
end

% Grafico las funciones de forma
figure                 % Creo un lienzo
grid on                % creo la rejilla
hold on;               % Para que no se sobreescriban los graficos
h = zeros(5,1);
color = ['r', 'b', 'c', 'm', 'k'];
for i = 1:5
    h(i) = fplot(N{i}, [-1 1], 'Color', color(i), 'LineWidth', 2);
end
title('Funciones de Forma Lagrangianas de CINCO nodos')
xlabel('\xi'); 
ylabel('N_i(\xi)');
plot([-1 -1/2 0 1/2 1],[0 0 0 0 0], 'ko', [-1 -1/2 0 1/2 1],[1 1 1 1 1], 'ko')
axis([-1 1 -0.6 1.2])
legend(h, 'N1(\xi)','N2(\xi)','N3(\xi)','N4(\xi)','N5(\xi)', 'Location','Best');

%% -------------------------------------------------------------------------
%% Calcular correctamente los polinomios de las funciones de forma 1D
function N = calc_N(xp, yp, var)
    % se ve verifican los tamanios de los vectores xp y yp
    nx = length(xp);
    ny = length(yp);
    assert(nx == ny, 'Los vectores xp y yp deben tener el mismo tamanio');

    % se calculan los coeficientes de los polinomios
    c = polyfit(xp, yp, nx-1);
    
    % se eliminan los errores en la aproximacion numerica, haciendo los
    % coeficientes demasiado pequenios igual a cero
    c(abs(c) < 1e-10) = 0;
    
    % con los coeficientes corregidos se calculan las funciones de forma
    N = poly2sym(c, var);
end
